### PR TITLE
Add multi-level chunk cache

### LINKED ISSUE #6240: Store Gateway: multi level chunks cache
**Is your feature request related to a problem? Please describe.**

Similar to the existing multi level cache for index, we want to support multi level cache for chunks.

**Describe the solution you'd like**

Depends on https://github.com/cortexproject/cortex/issues/6239.

It can be similar to https://github.com/cortexproject/cortex/blob/master/pkg/storage/tsdb/multilevel_cache.go but for caching bucket cache interface.

### PR DESCRIPTION
**Pull request by SungJin1212**
Support multi-level chunk cache like a multi-level cache for the index and add metrics for tracking multi-level cache behavior.

- `cortex_store_multilevel_chunks_cache_fetch_duration_seconds`, tracks latency to fetch item
- `cortex_store_multilevel_chunks_cache_backfill_duration_seconds`, tracks latency to backfill item
- `cortex_store_multilevel_chunks_cache_backfill_dropped_items_total`, tracks # of dropped items due to buffer fullness when backfilling
- `cortex_store_multilevel_chunks_cache_store_dropped_items_total`, tracks # of dropped items due to buffer fullness when storing

Add a multi level chunk cache
**Which issue(s) this PR fixes**:
Fixes #6240 

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by yeya24** [comment]
Can we fix tests so that we can merge?

**Comment by SungJin1212** [comment]
@yeya24 
I have fixed it.

**Comment by yeya24** [comment]
Hi @SungJin1212, I tested your PR locally but it got a panic when registering metrics.

My store gateway set up has metadata cache to use memcached and chunks cache to use multi level cache: inmemory as first level and memcached as second level.

I think the panic was caused by the new `level` label used by the chunks cache but the same metric in metadata cache doesn't have that. Prometheus registry expects the same label set across all metrics.

```
panic: a previously registered descriptor with the same fully-qualified name as Desc{fqName: "thanos_cache_memcached_requests_total", help: "Total number of items requests to memcached.", constLabels: {component="store-gateway",name="metadata-cache"}, variableLabels: {}} has different label names or a different help string

goroutine 1 [running]:
.../github.com/prometheus/client_golang/prometheus.(*wrappingRegisterer).MustRegister(0xc001413d70, {0xc001285800?, 0x0?, 0x0?})
```

**Comment by SungJin1212** [comment]
@yeya24 
Thank you for the review. I also reproduced it in my local. How about adding level label in metadata cache and just attach dummy label value like `tenancy.DefaultTenant`?

**Comment by SungJin1212** [comment]
@yeya24 
I removed a level label to chunk cache.

File: pkg/storage/tsdb/multilevel_chunk_cache.go
Comment (yeya24) [inline pkg/storage/tsdb/multilevel_chunk_cache.go:65]: Nit. Can we call it `chunks_cache` instead of `chunk_cache`? THen the naming is consistent with the config file.

File: pkg/storage/tsdb/multilevel_chunk_cache.go
Comment (yeya24) [inline pkg/storage/tsdb/multilevel_chunk_cache.go:50]: Let's reduce the backfill concurrency? 50 is a bit high. We should adjust the config for index cache as well in a separate PR.
2 ~ 4 might be a reasonable number. I have seen 50 goroutines doing backfilling taking a lot of CPUs.

**Review by yeya24 — COMMENTED** [review]
Thanks. The feature looks good. I have tried it in our test cluster and it saved quite a lot of Store Gateway and Chunks Cache bandwidth, especially when you have rules with a long lookback window.

File: pkg/storage/tsdb/multilevel_chunk_cache.go
Comment (SungJin1212) [reply pkg/storage/tsdb/multilevel_chunk_cache.go:50]: I pick the median. :)

**Review by yeya24 — APPROVED** [review]
Thanks. Waiting for another approval before merging this.

**Comment by yeya24** [comment]
We have a conflict seems aftering merging the previous PR

**Review by alanprot — APPROVED** [review]

**Comment by yeya24** [comment]
@SungJin1212 Need to fix lint

**Pull request merged**
